### ISSUE TITLE
ARC grows over zfs_arc_max (700M) and causes OOM panic

### ISSUE THREAD

**Issue by thegreatgazoo** [zfsonlinux/zfs#2840]
This is almost 100% reproducible with ZFS v0.6.3-1 on a test VM with 1.8G memory:
[root@eagle-44vm1 ~]# arcstat.py 2
    time  read  miss  miss%  dmis  dm%  pmis  pm%  mmis  mm%  arcsz     c
......
22:10:48  1.0K  1.0K     98    32   66   992  100     0    0   700M  700M
22:10:50   682   672     98    21   66   651  100     0    0   700M  700M
22:10:52   715   704     98    22   66   682  100     0    0   700M  700M
22:10:54   801   789     98    25   66   764  100     0    0   700M  700M
22:10:56   638   632     99    32   85   600  100    13   71   722M  700M
22:10:58   495   495    100    30  100   465  100    15  100   728M  700M
22:11:00   396   396    100    24  100   372  100    12  100   774M  700M
22:11:02   579   579    100    35  100   544  100    17  100   858M  700M
22:11:05   278   278    100    16  100   262  100     8  100   888M  700M
22:11:07   331   331    100    21  100   310  100    11  100   929M  700M
22:11:09   296   296     99    17   97   279  100     8  100   974M  700M
22:11:11   330   330    100    20  100   310  100    10  100  1017M  700M
22:11:13   264   264    100    16  100   248  100     8  100   1.0G  700M
22:11:15   330   330    100    20  100   310  100    10  100   1.1G  700M
22:11:17   278   278     99    18   97   260  100     9  100   1.1G  700M
22:11:19   231   228     98    14   84   214  100     7  100   1.1G  700M
22:11:21   279   274     98    18   78   256  100    10  100   1.2G  700M
22:11:23   278   276     99    15   88   261  100     7  100   1.2G  700M
22:11:25   827   827    100    52  100   775  100    26  100   1.4G  633M
22:11:27   314   314     99    40   98   274  100    29  100   1.4G  208M
22:11:29   625   621     99    54   92   566  100    37  100   1.5G  4.0M

At 22:10:54 the ARC size began to go out of control and grew over the 700M limit, and eventually caused an OOM panic.

At 22:11:23, the system seemed to begin trying to shrink the ARC. It seemed overly aggressive to set the ARC target size to only 4.0M. Also once the shrink began, there were still way too many prefetching requests (in fact, the majority of the request were prefetching). It seemed to me that once we began shrinking the ARC, it'd make sense to limit prefetching and leave precious ARC space to demand reads.

So in short there appeared to be 3 problems here:
1. ARC grows way over zfs_arc_max (to over 2*zfs_arc_max in the end).
2. ARC shrinking seemed too aggressive.
3. Should limit prefetching when shrinking the ARC.

Then I disabled prefetching and ran the test again (ARC increased to 800M, all else the same). The test completed successfully, and ARC size never grew over the limit. So the prefetching seemed to be the trouble maker here.

As this is very reproducible, if any debugging information is needed, please let me know.

**Comment by behlendorf** [comment]
Closing.  Since this issue was last updated multiple improvements have been made to the ARC to address this kind of issue.
